pkg/controller/master/node/node_down_controller.go

```diff
@@ -3,278 +3,315 @@ package node
-			pod.Name,
-			&metav1.DeleteOptions{
-				GracePeriodSeconds: &gracePeriod,
-			}); err != nil {
-			errs = multierr.Append(errs, fmt.Errorf("failed to delete pod %s/%s: %w", pod.Namespace, pod.Name, err))
-			continue
+	switch cond.Status {
+	case corev1.ConditionFalse:
+	case corev1.ConditionUnknown:
+		// check if node is reachable
+		nodes, err := h.nodeCache.List(labels.Everything())
+		if err != nil {
+			return err
 		}
-
-		names, err := h.getPVCsName(&pod)
+		targetIP := get_ip_from_other_nodes(node, nodes)
+		if targetIP == "" {
+			logrus.Warnf("can't get reachable IP from other nodes for node %s, skip this time", node.Name)
+			return nil
+		}
+		connected, err := pingNode(targetIP)
 		if err != nil {
-			errs = multierr.Append(errs, fmt.Errorf("failed to get PVCs name of pod %s/%s: %w", pod.Namespace, pod.Name, err))
-			continue
+			return fmt.Errorf("failed to ping node %s at ip %s: %w", node.Name, targetIP, err)
+		}
+		if connected {
+			waitEnough, err := h.tryWaitForKubelet(node)
+			if err != nil {
+				return err
+			}
+			if !waitEnough {
+				return nil
+			}
 		}
-		pvcSet.Insert(names...)
-	}
 
-	if err := h.deleteVolumeAttachments(node.Name, pvcSet); err != nil {
-		errs = multierr.Append(errs, fmt.Errorf("failed to delete VolumeAttachments: %w", err))
-	}
-	if errs != nil {
-		return node, errs
-	}
+		// we wait enough or node is unreachable, add taints to trigger migration
+		if getNodeTaint(node.Spec.Taints, kubevirtDrainTaintKey) == nil {
+			if err := h.triggerMigration(node); err != nil {
+				return fmt.Errorf("failed to trigger migration for node %s: %w", node.Name, err)
+			}
+			logrus.Infof("triggered migration for node %s", node.Name)
+			h.nodes.EnqueueAfter(node.Name, 5*time.Second)
+		}
 
-	return h.resetHeartbeat(node)
-}
+		if getNodeTaint(node.Spec.Taints, nonGracefulTaintKey) == nil {
+			if err := h.cleanupStuckResources(node); err != nil {
+				return fmt.Errorf("failed to cleanup stuck resources for node %s: %w", node.Name, err)
+			}
+			logrus.Infof("cleaned up stuck resources for node %s", node.Name)
+		}
 
-func (h *nodeDownHandler) OnVMForceResetPolicyChanged(_ string, setting *harvesterv1.Setting) (*harvesterv1.Setting, error) {
-	if setting == nil || setting.DeletionTimestamp != nil ||
-		setting.Name != settings.VMForceResetPolicySettingName || setting.Value == "" {
-		return setting, nil
+		return nil
+	case corev1.ConditionTrue:
+		// reset taint if node is healthy again
+		taintKeys := []string{}
+		if taint := getNodeTaint(node.Spec.Taints, kubevirtDrainTaintKey); taint != nil {
+			taintKeys = append(taintKeys, kubevirtDrainTaintKey)
+		}
+		if taint := getNodeTaint(node.Spec.Taints, nonGracefulTaintKey); taint != nil {
+			taintKeys = append(taintKeys, nonGracefulTaintKey)
+		}
+		if len(taintKeys) > 0 {
+			if err := h.removeTaints(node, taintKeys...); err != nil {
+				return fmt.Errorf("failed to remove taints %v from node %s: %w", taintKeys, node.Name, err)
+			}
+		}
```

Comment:
nit: we already take care the existence checks in removeTaints.
```suggestion
		if err := h.removeTaints(node, kubevirtDrainTaintKey, nonGracefulTaintKey); err != nil {
			return fmt.Errorf("failed to remove taints from node %s: %w", node.Name, err)
		}
```

Reply:
updated, thanks for the suggestion!